### PR TITLE
Extract operators from execution epic

### PR DESCRIPTION
**Pull request by rgbkrk**
This extracts some operators out of our execution epic for use elsewhere while cutting down the messaging logic from the desktop app. My hope is that the execution epic will only be tied to its core function of creating a notebook from the messages.

Generally I want to create some observable operators that you can use like this:

```js
iopub.pipe(
  childOf(someMessage),
  outputs()
)
```

Where the signature on `outputs` is `() => (Observable<JupyterMessage>): Observable<JupyterOutput>`

While doing this refactor I was pretty tempted to set this up quite differently by pumping the pure messages through the redux store instead, coupled with the cell ID. All the `takeUntil` logic would still be there, the execution epic would become a glorified router of messages that came from the parent ID to the proper cell ID.

------------------

Now that I've gone through this, we've got some nice primitives here which may make building kernels really easy as well. If you look at the tests you'll see the declarative nature of our message creators and how to work with our operators:

```js
import {of} from 'rxjs/observable/of';

import {
  createMessage,
  getUsername,
  createExecuteRequest,
  convertOutputMessageToNotebookFormat,
  outputs,
  payloads,
  executionCounts,
  executionStates,
} from '@nteract/messaging';

import {
  executeInput,
  displayData,
  updateDisplayData,
  executeResult,
  error,
  stream,
  status,
  executeReply,
  message,
  // TODO: Move outside this lib namespace
} from '@nteract/messaging/lib/messages';

// @format
of(
  status('starting'),
  status('idle'),
  status('busy'),
  executeInput({
    code: "display('woo')\ndisplay('hoo')",
    execution_count: 0,
  }),
  displayData({data: {'text/plain': 'woo'}}),
  displayData({data: {'text/plain': 'hoo'}}),
  status('idle'),
).pipe(executionCounts());
```

Since these are just JSON payloads and observable operators, people can still make the interfaces they want, mixing and matching styles as they see fit.

### CODE REVIEW

**Comment by rgbkrk** [comment]
~I completely broke all the tests as I suspected on this as I took out all the action specific functions that took observables and created operators.~ All set.

File: packages/desktop/test/renderer/epics/execute-spec.js [hunk 1]
```diff
@@ -89,139 +81,6 @@ describe("executeCellStream", () => {
...
-  it("creates an execute_request message", () => {
-    const code = 'print("test")';
-    const executeRequest = createExecuteRequest(code);
-
-    expect(executeRequest.content.code).to.equal(code);
-    expect(executeRequest.header.msg_type).to.equal("execute_request");
-  });
-});
-
-describe("msgSpecToNotebookFormat", () => {
```
Comment (lgeiger) [inline packages/desktop/test/renderer/epics/execute-spec.js:102]: Want to bring over the tests to the `@nteract/messaging`?

**Review by lgeiger — APPROVED** [review]
Thanks a lot for this refactor, this looks great!

File: packages/desktop/test/renderer/epics/execute-spec.js (hunk 1 above)
Comment (rgbkrk) [reply packages/desktop/test/renderer/epics/execute-spec.js:102]: Yeah I can do that and I think I'll clean up some comments that are in here too.

Comment (lgeiger) [reply packages/desktop/test/renderer/epics/execute-spec.js:102]: Awesome thanks 💯

File: packages/messaging/src/index.js [hunk 2]
```diff
@@ -37,6 +57,25 @@ export function createMessage(msg_type: string, fields: Object = {}) {
...
 
 /**
+ * Insert the content requisite for a code request to a kernel message.
+ *
+ * @param {String} code - Code to be executed in a message to the kernel.
+ * @return {Object} msg - Message object containing the code to be sent.
+ */
+export function createExecuteRequest(code: string) {
+  const executeRequest = createMessage("execute_request");
+  executeRequest.content = {
```
Comment (rgbkrk) [inline packages/messaging/src/index.js]: I'm a little confused about how this code isn't covered. I thought our messaging tests were running through the `jest` apparatus and that codecov was getting these. Any ideas @lgeiger?

**Comment by rgbkrk** [comment]
I'm now wishing that we had a few namespaces for this stuff:

```
// all our operators for working on streams of jupyter messages
import { outputs } from '@nteract/messaging/operators'

// Basically all the createXMessage functions, except as clean composable names
import { displayData } from '@nteract/messaging/messages'

of(
  executeInput({ execution_count: 1 }),
  executeReply(),
  displayData(c1),
  executeResult(c2),
  stream(c3),
  stream(c4)
).pipe(
  childOf(message),
  outputs()
)
```

The result of that would be (since filtered on outputs):

```
of(
  displayData(c1),
  executeResult(c2),
  stream(c3),
  stream(c4),
)
```

**Comment by rgbkrk** [comment]
Now that I've written that out, it must be done. I'm drooling at the declarative-ness.

File: packages/messaging/__tests__/messaging-spec.js [hunk 3]
```diff
@@ -188,7 +197,32 @@ describe("convertOutputMessageToNotebookFormat", () => {
 
 describe("outputs", () => {
   it("should be tested", () => {
-    expect(outputs()).toBeTruthy();
```
Comment (rgbkrk) [inline packages/messaging/__tests__/messaging-spec.js]: Ok, this is the commit you should all see. Look how clean and declarative this is. 😄

Comment (rgbkrk) [reply packages/messaging/__tests__/messaging-spec.js]: /cc @captainsafia @lgeiger

File: packages/desktop/src/notebook/epics/execute.js [hunk 4]
```diff
@@ -229,43 +81,58 @@ export function executeCellStream(
...
+      startWith(updateCellStatus(id, "busy"))
+    ),
+
     // Update the input numbering: `[ ]`
-    updateCellNumberingAction(id, cellMessages),
-    // Handle all nbformattable messages
-    handleFormattableMessages(id, cellMessages)
+    cellMessages.pipe(
+      executionCounts(),
+      map(ct => updateCellExecutionCount(id, ct))
```
Comment (rgbkrk) [inline packages/desktop/src/notebook/epics/execute.js:124]: How is this being reported as not being covered (it's a test that jest targets)?!

**Comment by rgbkrk** [comment]
Something seems a bit off with the coverage reports.

**Comment by rgbkrk** [comment]
This last commit should be it -- the only open questions for me are what namespaces we want. My first inclination is:

* `@nteract/messaging/operators`
* `@nteract/messaging/messages`

I could make them all just be part of the top level `@nteract/messaging` namespace too.

**Comment by rgbkrk** [comment]
Since this is working well and I'm ready to iterate on other parts that build on this, I'm going to go ahead and merge it. Thanks all!

**Review by lgeiger — COMMENTED** [review]
This is great! Thanks for the work 🎉

**Pull request merged**